experimental/cuda2/graph_command_buffer.c

```diff
@@ -708,6 +908,103 @@ static iree_status_t iree_hal_cuda2_graph_command_buffer_execute_commands(
+                    command_buffer, &segment->dispatch));
+      } break;
+      case IREE_HAL_CUDA_COMMAND_SEGMENT_ACTION_FILL_BUFFER: {
+        IREE_RETURN_AND_END_ZONE_IF_ERROR(
+            z0, iree_hal_cuda2_command_segment_record_fill_buffer(
+                    command_buffer, &segment->fill_buffer));
+      } break;
+      case IREE_HAL_CUDA_COMMAND_SEGMENT_ACTION_COPY_BUFFER: {
+        IREE_RETURN_AND_END_ZONE_IF_ERROR(
+            z0, iree_hal_cuda2_command_segment_record_copy_buffer(
+                    command_buffer, &segment->copy_buffer));
+      } break;
+      case IREE_HAL_CUDA_COMMAND_SEGMENT_ACTION_UPDATE_BUFFER: {
+        IREE_RETURN_AND_END_ZONE_IF_ERROR(
+            z0, iree_hal_cuda2_command_segment_record_update_buffer(
+                    command_buffer, &segment->update_buffer));
+      } break;
+      default:
+        IREE_ASSERT(false, "unhandled command segment kind");
+        break;
+    }
+  }
+
+  IREE_TRACE_ZONE_END(z0);
+  return iree_ok_status();
+}
+
+static iree_status_t iree_hal_cuda2_graph_command_buffer_begin(
+    iree_hal_command_buffer_t* base_command_buffer) {
+  iree_hal_cuda2_graph_command_buffer_t* command_buffer =
+      iree_hal_cuda2_graph_command_buffer_cast(base_command_buffer);
+
+  if (command_buffer->cu_graph != NULL) {
+    return iree_make_status(IREE_STATUS_FAILED_PRECONDITION,
+                            "cannot re-record command buffer");
+  }
+
+  iree_hal_cuda2_command_segment_list_reset(&command_buffer->segments);
+  iree_arena_reset(&command_buffer->arena);
+
+  return iree_ok_status();
+}
+
+static iree_status_t iree_hal_cuda2_graph_command_buffer_end(
+    iree_hal_command_buffer_t* base_command_buffer) {
+  iree_hal_cuda2_graph_command_buffer_t* command_buffer =
+      iree_hal_cuda2_graph_command_buffer_cast(base_command_buffer);
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  // Create a new empty graph to record into.
+  IREE_CUDA_RETURN_AND_END_ZONE_IF_ERROR(
+      z0, command_buffer->symbols,
+      cuGraphCreate(&command_buffer->cu_graph, /*flags=*/0), "cuGraphCreate");
+
+  // Reset state used during recording.
+  command_buffer->last_node = NULL;
+
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0, iree_hal_cuda2_command_segment_record(command_buffer));
+
+  // Compile the graph.
+  CUgraphNode error_node = NULL;
+  iree_status_t status = IREE_CURESULT_TO_STATUS(
+      command_buffer->symbols,
+      cuGraphInstantiate(&command_buffer->cu_graph_exec,
+                         command_buffer->cu_graph, &error_node,
+                         /*logBuffer=*/NULL, /*bufferSize=*/0));
+  if (iree_status_is_ok(status)) {
+    // No longer need the source graph used for construction.
+    IREE_CUDA_IGNORE_ERROR(command_buffer->symbols,
+                           cuGraphDestroy(command_buffer->cu_graph));
+    command_buffer->cu_graph = NULL;
+  }
+
+  iree_hal_resource_set_freeze(command_buffer->resource_set);
+
+  IREE_TRACE_ZONE_END(z0);
+  return iree_ok_status();
```

Comment:
I think that the biggest performance problem will actually be from a call to `cuGraphInstantiate` on every `iree_hal_cuda2_graph_command_buffer_end`, NVIDIA recommends keeping executable around and updating them with new parameters instead of destroy/instantiate cycle.